pkg/kubeapply/kubeapply.go

```diff
@@ -110,63 +110,56 @@ func (collection YAMLCollection) ApplyAndWait(
 		deadline := time.Now().Add(perPhaseTimeout)
 		err := applyAndWait(ctx, kubeinfo, deadline, debug, dryRun, collection[phaseName])
 		if err != nil {
-			if err == errorDeadlineExceeded {
-				err = errors.Errorf("phase %q not ready after %v", phaseName, perPhaseTimeout)
+			if errors.Is(err, errorDeadlineExceeded) {
+				err = fmt.Errorf("phase %q not ready after %v: %w", phaseName, perPhaseTimeout, err)
 			}
 			return err
 		}
 	}
 	return nil
 }
 
-func applyAndWait(ctx context.Context, kubeinfo *k8s.KubeInfo, deadline time.Time, debug, dryRun bool, filenames []string) error {
-	expanded, err := expand(ctx, filenames)
+func applyAndWait(ctx context.Context, kubeinfo *k8s.KubeInfo, deadline time.Time, debug, dryRun bool, sourceFilenames []string) error {
+	expandedFilenames, err := expand(ctx, sourceFilenames)
 	if err != nil {
-		return err
+		return fmt.Errorf("expanding YAML: %w", err)
 	}
 
 	cli, err := k8s.NewClient(kubeinfo)
 	if err != nil {
-		return errors.Wrapf(err, "kubeapply: error connecting to cluster %v", kubeinfo)
+		return fmt.Errorf("connecting to cluster %v: %w", kubeinfo, err)
 	}
 	waiter, err := NewWaiter(cli.Watcher())
 	if err != nil {
 		return err
 	}
 
 	valid := make(map[string]bool)
-	var msgs []string
-	for _, n := range expanded {
-		err := waiter.Scan(ctx, n)
-		if err != nil {
-			msgs = append(msgs, fmt.Sprintf("%s: %v\n", n, err))
-			valid[n] = false
-		} else {
-			valid[n] = true
+	var scanErrs derror.MultiError
+	for _, filename := range expandedFilenames {
+		valid[filename] = true
+		if err := waiter.Scan(ctx, filename); err != nil {
+			scanErrs = append(scanErrs, fmt.Errorf("watch %q: %w", filename, err))
+			valid[filename] = false
 		}
 	}
-
-	if len(msgs) == 0 {
-		err = kubectlApply(ctx, kubeinfo, dryRun, expanded)
-	}
-
 	if !debug {
-		for _, n := range expanded {
-			if valid[n] {
-				err := os.Remove(n)
-				if err != nil {
-					dlog.Print(ctx, err)
+		defer func() {
```

Comment:
Probably worth a comment about intentions here. (Admittedly, some of my tripping-up here might be from reading the _diff_ rather than the source, and wondering what the point of adding the deferral was.)

Reply:
I've added a comment, but I think a significant part of tripping up is looking at the diff.